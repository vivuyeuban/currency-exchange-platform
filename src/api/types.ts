export type TRateRequestParams = {
    amount: string | number;
    from: string;
    to: string;
};

export type TRateResponse = {
    rate: number | undefined;
    amount: number | undefined;
} | undefined;

export type TOriginalResponse = {
    info: {
        rate: number;
    };
    result: number;
} | undefined;

export type TMockApi = {
    isError: boolean;
    isSuccess: boolean;
    data: TRateResponse;
    refetch: () => void;
};
