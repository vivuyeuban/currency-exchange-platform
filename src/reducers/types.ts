export type TAccount = {
    id: string;
    name: string;
    currency: string;
    balance: number;
};

export type TAccountsState = {
    accounts: {
        [key: string]: TAccount;
    };
};
