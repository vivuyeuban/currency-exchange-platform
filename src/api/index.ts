import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { API_BASE_URL } from '../constants';
import { TRateResponse, TOriginalResponse } from './types';

// Define a service using a base URL and expected endpoints
export const conversionApi = createApi({
  reducerPath: 'conversion',
  baseQuery: fetchBaseQuery({ baseUrl: API_BASE_URL }),
  endpoints: (builder) => ({
    getConversionRates: builder.query({
      query: ({ amount, from, to }) => `convert?from=${from}&to=${to}&amount=${amount}`,
      transformResponse: (response: TOriginalResponse): TRateResponse => ({
        rate: response?.info?.rate,
        amount: response?.result,
      }),
    }),
  }),
});

// auto-generated based on the defined endpoints
export const { useGetConversionRatesQuery } = conversionApi;
