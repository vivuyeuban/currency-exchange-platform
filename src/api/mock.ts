import React from 'react';
import { TMockApi } from './types';

// eslint-disable-next-line
const getUseMockApi = (success = true): CallableFunction => (_props: any, _options: any): TMockApi => {
  const [data, setData] = React.useState({ rate: 1.23, amount: 234.34 });

  const refetch = () => {
    setData({
      rate: Math.random() * 2 + 1,
      amount: 100 * Math.random() * 20 + 1,
    });
  };
  return {
    isError: !success,
    isSuccess: success,
    data: success ? data : undefined,
    refetch,
  };
};

export default getUseMockApi;
