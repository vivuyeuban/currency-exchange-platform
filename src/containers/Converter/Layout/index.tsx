import React from 'react';

import FormHelperText from '@material-ui/core/FormHelperText';
import ImportExportIcon from '@material-ui/icons/ImportExport';
import Divider from '@material-ui/core/Divider';
import Typography from '@material-ui/core/Typography';

import AccountSelect from '../../../components/AccountSelect';
import CurrencyInput from '../../../components/CurrencyInput';
import TopToast from '../../../components/TopToast';
import Button from '../../../components/Button';

import { useStyles } from './styles';
import { getCurrencySymbol } from '../../../utils/currencies';

import type { TConverterLayoutProps } from './types';

const ConverterLayout = React.memo(({
  showToast = false,
  onToastClose,
  accounts = [],
  src,
  dest,
  isValid = false,
  exchangeRate,
  onSumbit,
  onChangeSrcToDest,
}: TConverterLayoutProps) => {
  const classes = useStyles();

  const outOfBound = +src.balance < +src.amount;
  return (
    <div>
      <TopToast
        severity="success"
        open={showToast}
        onDelayedClose={onToastClose}
      >
        Done!
      </TopToast>
      <div className={classes.container}>
        <Typography
          className={classes.customTypographyStyle}
          variant="h5"
          display="block"
          gutterBottom
        >
          I want to exchange
        </Typography>
        <AccountSelect
          id="sourceAcc"
          name="sourceAcc"
          accounts={accounts}
          value={src.accountId}
          className={classes.select}
          label="From"
          onChange={src.onAccountChange}
        />
        <Button
          className={classes.customChangeStyle}
          onClick={onChangeSrcToDest}
        >
          <ImportExportIcon />
        </Button>
        <AccountSelect
          id="destAcc"
          name="destAcc"
          value={dest.accountId}
          accounts={accounts}
          className={classes.select}
          label="To"
          onChange={dest.onAccountChange}
        />
        {
          !outOfBound
          && !!src.accountId
          && !!dest.accountId
          && !!src.currency
          && !!dest.currency
          && src.currency !== dest.currency
          && (
            <>
              <FormHelperText className={classes.rateInfo}>
                {!exchangeRate ? 'Loading...' : `Rate: ${getCurrencySymbol(src.currency)}1 = ${getCurrencySymbol(dest.currency)}${exchangeRate}`}
              </FormHelperText>
              <Divider variant="middle" />
            </>
          )
        }
        <div className={classes.row}>
          <CurrencyInput
            id="srcAmount"
            name="srcAmount"
            error={outOfBound}
            disabled={!src.accountId}
            helperText={outOfBound && 'Not enough money'}
            className={classes.marginRight}
            currency={src.currency}
            value={src.amount}
            onChange={src.onValChange}
            label="Amount"
          />
          <CurrencyInput
            id="destAmount"
            name="destAmount"
            label="Result"
            disabled={!dest.accountId}
            currency={dest.currency}
            value={dest.amount}
            onChange={dest.onValChange}
          />
        </div>
        <Button
          variant="contained"
          className={classes.customExachangeButtonStyle}
          disabled={!isValid}
          onClick={onSumbit}
        >
          Exchange
        </Button>
      </div>
    </div>
  );
});

export default ConverterLayout;
