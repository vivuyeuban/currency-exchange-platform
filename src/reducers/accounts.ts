/* eslint no-param-reassign: ["error", { "props": false }] */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import type { TAccount, TAccountsState } from './types';

const initialState: TAccountsState = {
  accounts: {},
};

export const accountsSlice = createSlice({
  name: 'acoounts',
  initialState,
  reducers: {
    updateAccountBalance: (state, action: PayloadAction<{ id: TAccount['id'], newBalance: TAccount['balance'] }>) => {
      state.accounts[action.payload.id].balance = action.payload.newBalance;
    },
    setAccounts: (state, action: PayloadAction<Array<TAccount>>) => {
      action.payload.forEach((acc) => {
        state.accounts[acc.id] = acc;
      });
    },
    removeAccount: (state, action: PayloadAction<TAccount['id']>) => {
      delete state.accounts[action.payload];
    },
    clearAccounts: (state) => {
      state.accounts = {};
    },
  },
});

export const {
  updateAccountBalance,
  setAccounts,
  removeAccount,
  clearAccounts,
} = accountsSlice.actions;
