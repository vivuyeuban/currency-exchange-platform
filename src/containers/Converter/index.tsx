import React from 'react';

import ConverterLayout from './Layout';
import useConversionForm from '../../hooks/useConversionForm';
import useAccounts from '../../hooks/useAccounts';

import { useGetConversionRatesQuery } from '../../api';

import type { TAccountSelectOnChangeProps } from '../../components/AccountSelect/types';

const Converter = React.memo(() => {
  const [showSuccessToast, setShowSuccessToast] = React.useState(false);
  const {
    state,
    list,
    getById,
  } = useAccounts();

  const {
    sourceAcc,
    sourceAmount,
    destAcc,
    destAmount,
    setSourceAcc,
    setDestAcc,
    setSrcAmount,
    setDestAmount,
    isValid,
    exchangeRate,
    onSubmit,
    onChangeSrcToDest,
  } = useConversionForm({ accounts: state.accounts, useApiHook: useGetConversionRatesQuery });

  const onSourceAccChange = (event: TAccountSelectOnChangeProps) => {
    setSourceAcc(event.target.value);
  };

  const onDestAccChange = (event: TAccountSelectOnChangeProps) => {
    setDestAcc(event.target.value);
  };

  const onExchange = () => {
    onSubmit();
    setShowSuccessToast(true);
  };

  const onExchangeSrcToDestValue = () => {
    onChangeSrcToDest();
  };

  const hideToast = React.useCallback(() => {
    setShowSuccessToast(false);
  }, []);

  return (
    <ConverterLayout
      accounts={list}
      isValid={isValid}
      showToast={showSuccessToast}
      onToastClose={hideToast}
      onSumbit={onExchange}
      onChangeSrcToDest={onExchangeSrcToDestValue}
      exchangeRate={exchangeRate}
      src={{
        accountId: sourceAcc,
        currency: getById(sourceAcc)?.currency,
        amount: sourceAmount,
        balance: getById(sourceAcc)?.balance,
        onAccountChange: onSourceAccChange,
        onValChange: setSrcAmount,
      }}
      dest={{
        accountId: destAcc,
        currency: getById(destAcc)?.currency,
        amount: destAmount,
        onAccountChange: onDestAccChange,
        onValChange: setDestAmount,
      }}
    />
  );
});

export default Converter;
