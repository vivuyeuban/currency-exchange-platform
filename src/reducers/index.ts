import { combineReducers } from '@reduxjs/toolkit';

import { conversionApi } from '../api';
import { accountsSlice } from './accounts';

export const rootReducer = combineReducers({
  accounts: accountsSlice.reducer,
  [conversionApi.reducerPath]: conversionApi.reducer,
});
