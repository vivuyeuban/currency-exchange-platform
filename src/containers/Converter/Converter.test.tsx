import React from 'react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { rootReducer } from '../../reducers';
import { conversionApi } from '../../api';

import Converter from './index';

const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(conversionApi.middleware),
});

const { container } = render(
  <Provider store={store}>
    <Converter />
  </Provider>,
);

test('matches the snapshot', () => {
  expect(container).toMatchSnapshot();
});
