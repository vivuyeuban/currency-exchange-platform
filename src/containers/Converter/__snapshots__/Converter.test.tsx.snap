// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<div>
  <div>
    <div
      class="MuiPaper-root MuiAlert-root MuiAlert-filledSuccess makeStyles-root-9  MuiPaper-elevation0"
      role="alert"
    >
      <div
        class="MuiAlert-icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4C12.76,4 13.5,4.11 14.2, 4.31L15.77,2.74C14.61,2.26 13.34,2 12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0, 0 22,12M7.91,10.08L6.5,11.5L11,16L21,6L19.59,4.58L11,13.17L7.91,10.08Z"
          />
        </svg>
      </div>
      <div
        class="MuiAlert-message"
      >
        Done!
      </div>
    </div>
    <div
      class="makeStyles-container-4"
    >
      <h5
        class="MuiTypography-root makeStyles-customTypographyStyle-8 MuiTypography-h5 MuiTypography-gutterBottom MuiTypography-displayBlock"
      >
        I want to exchange
      </h5>
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-select-2"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="sourceAcc"
          id="sourceAcc-label"
        >
          From
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="sourceAcc-label sourceAcc"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
            id="sourceAcc"
            role="button"
            tabindex="0"
          >
            <span>
              ​
            </span>
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            name="sourceAcc"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-13"
            >
              <span>
                From
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-customChangeStyle-7 MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M9 3L5 6.99h3V14h2V6.99h3L9 3zm7 14.01V10h-2v7.01h-3L15 21l4-3.99h-3z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-select-2"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="destAcc"
          id="destAcc-label"
        >
          To
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="destAcc-label destAcc"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
            id="destAcc"
            role="button"
            tabindex="0"
          >
            <span>
              ​
            </span>
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            name="destAcc"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-13"
            >
              <span>
                To
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="makeStyles-row-3"
      >
        <div
          class="MuiFormControl-root makeStyles-marginRight-1"
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-disabled Mui-disabled"
              data-shrink="true"
              for="srcAmount"
              id="srcAmount-label"
            >
              Amount
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart"
            >
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionStart"
                data-testid="currency-sign"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                >
                  $
                </p>
              </div>
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart"
                disabled=""
                id="srcAmount"
                name="srcAmount"
                placeholder="0.00"
                style="text-align: right;"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-13 PrivateNotchedOutline-legendNotched-14"
                >
                  <span>
                    Amount
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiFormControl-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-disabled Mui-disabled"
              data-shrink="true"
              for="destAmount"
              id="destAmount-label"
            >
              Result
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart"
            >
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionStart"
                data-testid="currency-sign"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                >
                  $
                </p>
              </div>
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart"
                disabled=""
                id="destAmount"
                name="destAmount"
                placeholder="0.00"
                style="text-align: right;"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-13 PrivateNotchedOutline-legendNotched-14"
                >
                  <span>
                    Result
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-customExachangeButtonStyle-6 MuiButton-containedPrimary Mui-disabled Mui-disabled"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Exchange
        </span>
      </button>
    </div>
  </div>
</div>
`;
